packages/astro/src/vite-plugin-astro/hmr.ts

```diff
@@ -68,25 +58,39 @@ export async function handleHotUpdate(
 	}
 }
 
-function isStyleOnlyChanged(oldResult: CompileResult, newResult: CompileResult) {
-	return (
-		normalizeCode(oldResult.code) === normalizeCode(newResult.code) &&
-		// If style tags are added/removed, we need to regenerate the main Astro file
-		// so that its CSS imports are also added/removed
-		oldResult.css.length === newResult.css.length &&
-		!isArrayEqual(oldResult.css, newResult.css)
-	);
-}
+const frontmatterRE = /^\-\-\-.*?^\-\-\-/ms;
+const scriptRE = /<script(?:\s.*?)?>.*?<\/script>/gs;
+const styleRE = /<style(?:\s.*?)?>.*?<\/style>/gs;
+
+function isStyleOnlyChanged(oldCode: string, newCode: string) {
+	if (oldCode === newCode) return false;
+
+	// Before we can regex-capture style tags, we remove the frontmatter and scripts
+	// first as they could contain false-positive style tag matches. At the same time,
+	// we can also compare if they have changed and early out.
+
+	// Strip off and compare frontmatter
+	let oldFrontmatter = '';
+	let newFrontmatter = '';
+	oldCode = oldCode.replace(frontmatterRE, (m) => ((oldFrontmatter = m), ''));
+	newCode = newCode.replace(frontmatterRE, (m) => ((newFrontmatter = m), ''));
+	if (oldFrontmatter !== newFrontmatter) return false;
 
-const astroStyleImportRE = /import\s*"[^"]+astro&type=style[^"]+";/g;
-const sourceMappingUrlRE = /\/\/# sourceMappingURL=[^ ]+$/gm;
+	// Strip off and compare scripts
+	const oldScripts: string[] = [];
+	const newScripts: string[] = [];
+	oldCode = oldCode.replace(scriptRE, (m) => (oldScripts.push(m), ''));
+	newCode = newCode.replace(scriptRE, (m) => (newScripts.push(m), ''));
```

Comment:
`scriptRE` is a single global RegExp outside of the `isStyleChanged()` scope. Multiple executions of `isStyleChanged()` will continue from the previous `lastIndex`.

You need to manually reset `scriptRE.lastIndex = 0` at the top of `isStyleChanged` so the regex matches from the beginning of the file.

Reply:
I think `replace` will auto-reset it to 0, usually when it can't find anymore replacement, it'll set to 0 internally. I made a test to confirm this:

```js
const re = /a/g
const str = 'abc_abc_abc'.replace(re, () => '')
console.log(re.lastIndex) // 0
```

Reply:
> Because `@@replace` would keep calling `exec()` until it returns null, and `exec()` would automatically reset the regex's [lastIndex](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/lastIndex) to 0 when the last match fails, `@@replace` would typically not have side effects when it exits. However, when the regex is [sticky](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky) but not global, `lastIndex` would not be reset. In this case, each call to `replace()` may return a different result.
> - [MDN `@@replace`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/@@replace)

TIL. The nuances of regex continue to be wild.